crates/wasmtime/src/runtime/vm/instance/allocator/pooling/decommit_queue.rs

```diff
@@ -0,0 +1,177 @@
+//! A queue for batching decommits together.
+//!
+//! We don't immediately decommit a Wasm table/memory/stack/etc... eagerly, but
+//! instead batch them up to be decommitted together. This module implements
+//! that queuing and batching.
+//!
+//! Even when batching is "disabled" we still use this queue. Batching is
+//! disabled by specifying a batch size of one, in which case, this queue will
+//! immediately get flushed everytime we push onto it.
+
+use super::PoolingInstanceAllocator;
+use crate::vm::{MemoryAllocationIndex, MemoryImageSlot, Table, TableAllocationIndex};
+use alloc::vec::Vec;
+use core::mem;
+use smallvec::SmallVec;
+
+#[cfg(feature = "async")]
+use wasmtime_fiber::FiberStack;
+
+pub struct DecommitQueue {
+    batch_size: usize,
+    raw: SmallVec<[libc::iovec; 2]>,
+    memories: SmallVec<[(MemoryAllocationIndex, MemoryImageSlot); 1]>,
+    tables: SmallVec<[(TableAllocationIndex, Table); 1]>,
+    #[cfg(feature = "async")]
+    stacks: SmallVec<[FiberStack; 1]>,
+    //
+    // TODO: GC heaps are not well-integrated with the pooling allocator
+    // yet. Once we better integrate them, we should start (optionally) zeroing
+    // them, and batching that up here.
+    //
+    // #[cfg(feature = "gc")]
+    // pub gc_heaps: SmallVec<[(GcHeapAllocationIndex, Box<dyn GcHeap>); 1]>,
+}
+
+unsafe impl Send for DecommitQueue {}
+unsafe impl Sync for DecommitQueue {}
+
+impl std::fmt::Debug for DecommitQueue {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("DecommitQueue")
+            .field("batch_size", &self.batch_size)
+            .field(
+                "raw",
+                &self
+                    .raw
+                    .iter()
+                    .map(|v| (v.iov_base, v.iov_len))
+                    .collect::<Vec<_>>(),
+            )
+            .field("memories", &"..")
+            .field("tables", &"..")
+            .field("stacks", &"..")
+            .finish()
+    }
+}
+
+impl DecommitQueue {
+    /// Construct a new queue with the given batch size.
+    pub fn new(batch_size: u32) -> Self {
+        // Don't try to `madvise` more than the kernel can put in a single
+        // iovec-array at the same time.
+        let iov_max = u32::try_from(unsafe { libc::sysconf(libc::_SC_IOV_MAX) }).unwrap();
+        let batch_size = core::cmp::min(batch_size, iov_max);
+        let batch_size = usize::try_from(batch_size).unwrap();
+
+        Self {
+            batch_size,
+            raw: Default::default(),
+            memories: Default::default(),
+            tables: Default::default(),
+            #[cfg(feature = "async")]
+            stacks: Default::default(),
+        }
+    }
+
+    /// Take this queue and leave behind an empty queue with the same batch
+    /// size.
+    pub fn take(&mut self) -> Self {
+        let new = Self::new(u32::try_from(self.batch_size).unwrap());
```

Comment:
On glibc, `sysconf(_SC_IOV_MAX)` returns a compile-time constant, so I don't think we should worry about its speed: https://github.com/bminor/glibc/blob/dd5f891c1ad9f1b43b9db93afe2a55cbb7a6194e/sysdeps/posix/sysconf.c#L412-L418

Reply:
We have to handle the case where the batch size is larger than `IOV_MAX` anyways, since a single wasm memory can push two memory regions to decommit and so we can always blow our target batch size. Therefore, I think we should just remove this bit, and allow the batch decommit to break things up into smaller chunks if it must, since it has to have that logic anyways.